File: services/api/src/NATS/subscriptions/ai-interaction-subjects.ts
'use strict'

import chalk from 'chalk'

import NATS_Service from '@lixpi/nats-service'
import { log, info, infoStr, warn, err } from '@lixpi/debug-tools'
import { NATS_SUBJECTS, type AiModelId, type AiInteractionChatSendMessagePayload, type AiInteractionChatStopMessagePayload } from '@lixpi/constants'

import AiModel from '../../models/ai-model.ts'

const { AI_INTERACTION_SUBJECTS } = NATS_SUBJECTS

export const aiInteractionSubjects = [
    {
        subject: AI_INTERACTION_SUBJECTS.CHAT_SEND_MESSAGE,
        type: 'subscribe',
        queue: 'aiInteraction',
        payloadType: 'json',
        permissions: {
            pub: {
                allow: [
                    AI_INTERACTION_SUBJECTS.CHAT_SEND_MESSAGE
                ]
            },
            sub: {
                allow: [
                    `${AI_INTERACTION_SUBJECTS.CHAT_SEND_MESSAGE_RESPONSE}.>`
                ]
            }
        },
        handler: async (data, msg) => {
            const {
                user: {
                    userId,
                    stripeCustomerId
                },
                messages,
                aiModel,
                workspaceId,
                aiChatThreadId,
                organizationId
            } = data as {
                user: { userId: string; stripeCustomerId: string }
                workspaceId: string
                aiChatThreadId: string
                organizationId: string
            } & AiInteractionChatSendMessagePayload

            const [provider, model] = (aiModel as string).split(':')
            const natsService = await NATS_Service.getInstance();

            try {
                // Fetch AI model meta info with pricing (for llm-api service)
                const aiModelMetaInfo = await AiModel.getAiModel({ provider, model, omitPricing: false })

                if (!aiModelMetaInfo || !aiModelMetaInfo.modelVersion) {
                    err('AI model meta info not found in the database', { aiModel })

                    // Publish error directly to client
                    natsService.publish(`${AI_INTERACTION_SUBJECTS.CHAT_SEND_MESSAGE_RESPONSE}.${workspaceId}.${aiChatThreadId}`, {
                        error: `AI model not found: ${aiModel}`
                    })
                    return
                }

                // One stream per thread - use workspaceId:aiChatThreadId as unique key
                const instanceKey = `${workspaceId}:${aiChatThreadId}`

                infoStr([
                    chalk.cyan('🚀 [AI_INTERACTION] GATEWAY'),
                    ' :: Forwarding to llm-api',
                    ' :: instanceKey:',
                    chalk.yellow(instanceKey),
                    ' :: provider:',
                    chalk.green(provider)
                ])

                // Forward request to llm-api service via internal NATS subject
                natsService.publish(AI_INTERACTION_SUBJECTS.CHAT_PROCESS, {
                    messages,
                    aiModelMetaInfo,
                    workspaceId,
                    aiChatThreadId,
                    eventMeta: {
                        userId,
                        stripeCustomerId,
                        organizationId,
                        workspaceId,
                        aiChatThreadId
                    }
                })

                infoStr([
                    chalk.green('✅ [AI_INTERACTION] GATEWAY'),
                    ' :: Request forwarded to llm-api',
                    ' :: instanceKey:',
                    chalk.yellow(instanceKey)
                ])

            } catch (error) {
                err('❌ [AI_INTERACTION] GATEWAY ERROR:', error)

                // Publish error to client
                natsService.publish(`${AI_INTERACTION_SUBJECTS.CHAT_SEND_MESSAGE_RESPONSE}.${workspaceId}.${aiChatThreadId}`, {
                    error: error instanceof Error ? error.message : String(error)
                })
            }
        }
    },

    // Stop AI message streaming
    {
        subject: AI_INTERACTION_SUBJECTS.CHAT_STOP_MESSAGE,
        type: 'subscribe',
        queue: 'aiInteraction',
        payloadType: 'json',
        permissions: {
            pub: {
                allow: [
                    AI_INTERACTION_SUBJECTS.CHAT_STOP_MESSAGE
                ]
            }
        },
        handler: async (data, msg) => {
            const {
                user: {
                    userId
                },
                workspaceId,
                aiChatThreadId
            } = data as {
                user: { userId: string }
                workspaceId: string
                aiChatThreadId: string
            }

            const natsService = await NATS_Service.getInstance()
            const instanceKey = `${workspaceId}:${aiChatThreadId}`

            infoStr([
                chalk.yellow('🛑 [AI_INTERACTION] GATEWAY'),
                ' :: Relaying STOP to llm-api',
                ' :: instanceKey:',
                chalk.red(instanceKey)
            ])

            // Relay stop request to llm-api service
            natsService.publish(`${AI_INTERACTION_SUBJECTS.CHAT_STOP}.${instanceKey}`, {
                workspaceId,
                aiChatThreadId,
                userId
            })

            infoStr([
                chalk.green('✅ [AI_INTERACTION] GATEWAY'),
                ' :: STOP request relayed to llm-api',
                ' :: instanceKey:',
                chalk.yellow(instanceKey)
            ])
        }
    },

    // Handle errors from llm-api service
    {
        subject: `${AI_INTERACTION_SUBJECTS.CHAT_ERROR}.>`,
        type: 'subscribe',
        queue: 'aiInteraction',
        payloadType: 'json',
        permissions: {
            sub: {
                allow: [
                    `${AI_INTERACTION_SUBJECTS.CHAT_ERROR}.>`
                ]
            }
        },
        handler: async (data, msg) => {
            const { error, instanceKey } = data as { error: string; instanceKey: string }

            const natsService = await NATS_Service.getInstance()

            // Extract workspaceId and aiChatThreadId from instanceKey (format: workspaceId:aiChatThreadId)
            const [workspaceId, aiChatThreadId] = instanceKey.split(':')

            err('❌ [AI_INTERACTION] ERROR from llm-api:', { instanceKey, error })

            // Forward error to client
            natsService.publish(`${AI_INTERACTION_SUBJECTS.CHAT_SEND_MESSAGE_RESPONSE}.${workspaceId}.${aiChatThreadId}`, {
                error
            })
        }
    },
]

File: services/llm-api/src/NATS/subscriptions/ai_interaction_subjects.py
"""
AI Interaction NATS subscriptions for LLM API service.
Handles chat processing and stop requests.
"""

from lixpi_debug_tools import info, warn, err
from lixpi_constants import NATS_SUBJECTS

# Extract AI interaction subjects
AI_INTERACTION_SUBJECTS = NATS_SUBJECTS["AI_INTERACTION_SUBJECTS"]
CHAT_PROCESS = AI_INTERACTION_SUBJECTS["CHAT_PROCESS"]
CHAT_STOP = AI_INTERACTION_SUBJECTS["CHAT_STOP"]
CHAT_ERROR = AI_INTERACTION_SUBJECTS["CHAT_ERROR"]


def get_ai_interaction_subjects(registry):
    """
    Get AI interaction NATS subscription definitions.

    Args:
        registry: Provider registry instance for handler access

    Returns:
        List of subscription configurations
    """

    # Handler: Process chat requests from api service
    async def _handler_chat_process(data, msg):
        try:
            # Extract request data
            workspace_id = data.get('workspaceId')
            ai_chat_thread_id = data.get('aiChatThreadId')
            ai_model_meta_info = data.get('aiModelMetaInfo', {})
            provider_name = ai_model_meta_info.get('provider')

            if not workspace_id:
                err("Missing workspaceId in request")
                return

            if not ai_chat_thread_id:
                err("Missing aiChatThreadId in request")
                return

            if not provider_name:
                err("Missing provider in aiModelMetaInfo")
                return

            # Create instance key using workspaceId:aiChatThreadId
            instance_key = f"{workspace_id}:{ai_chat_thread_id}"

            info(f"Processing chat request for {instance_key} using {provider_name}")

            # Get or create provider instance
            provider = registry._get_or_create_instance(instance_key, provider_name)

            # Process the request through LangGraph workflow
            await provider.process(data)

            # Remove instance after completion
            registry._remove_instance(instance_key)

        except Exception as e:
            err(f"Error handling chat process: {e}")

            # Publish error back to services/api
            workspace_id = data.get('workspaceId', 'unknown')
            ai_chat_thread_id = data.get('aiChatThreadId', 'unknown')
            instance_key = f"{workspace_id}:{ai_chat_thread_id}"
            registry.nats_client.publish(
                f"{CHAT_ERROR}.{instance_key}",
                {
                    'error': str(e),
                    'instanceKey': instance_key
                }
            )

    # Handler: Stop streaming request
    async def _handler_chat_stop(data, msg):
        try:
            # Extract instance key from subject (ai.interaction.chat.stop.{instanceKey})
            instance_key = '.'.join(msg.subject.split('.')[4:])

            info(f"Received stop request for {instance_key}")

            # Find and stop the instance
            provider = registry.instances.get(instance_key)
            if provider:
                await provider.stop()
                info(f"Stopped instance: {instance_key}")
            else:
                warn(f"Instance not found: {instance_key}")

        except Exception as e:
            err(f"Error handling chat stop: {e}")

    return [
        # Process chat requests from api service
        {
            'subject': CHAT_PROCESS,
            'type': 'subscribe',
            'payloadType': 'json',
            'queue': 'llm-workers',
            'permissions': {
                'pub': {
                    'allow': [
                        f"{CHAT_ERROR}.>",  # Publish errors back to API
                        "ai.interaction.chat.receiveMessage.>"  # Stream LLM responses to web-ui
                    ]
                },
                'sub': {
                    'allow': [
                        CHAT_PROCESS,  # Subscribe to chat processing requests
                        f"{CHAT_STOP}.>"  # Subscribe to stop requests
                    ]
                }
            },
            'handler': _handler_chat_process
        },

        # Handle stop requests
        {
            'subject': f"{CHAT_STOP}.>",
            'type': 'subscribe',
            'payloadType': 'json',
            'permissions': {
                'sub': {
                    'allow': [
                        f"{CHAT_STOP}.>"
                    ]
                }
            },
            'handler': _handler_chat_stop
        }
    ]
